pkg/controller/node/node_controller.go

```diff
@@ -128,3 +133,20 @@ func (n *NodeController) generateAnnotation(nodeEgressIPConfigs []*cloudprovider
 	}
 	return string(serialized), nil
 }
+
+// isNodeInitializedByCloudControllerManager checks if this node was already initialized by the Cloud Controller
+// Manager. For further details, see https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/.
+func isNodeInitializedByCloudControllerManager(node *corev1.Node) bool {
+	// Nil input is not initialized.
+	if node == nil {
+		return false
+	}
+	// Look for the well known taint. If it can be found, then the node is not initialized.
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == cloudproviderapi.TaintExternalCloudProvider {
```

Comment:
@andreaskaris does it matter what Effect is associated with the taint? I'm guessing it does not.

Reply:
I thought about that, too, but I don't see a reason why anyone would mess with exactly that taint when the 3 valid values are:
~~~
TaintEffect = "NoSchedule"
TaintEffect = "PreferNoSchedule"
TaintEffect = "NoExecute"
~~~

I was about to push this amendment here but looking at the unit test, it felt a bit creepy to me:
~~~
diff --git a/pkg/controller/node/node_controller.go b/pkg/controller/node/node_controller.go
index 5b5773b..6dd4e63 100644
--- a/pkg/controller/node/node_controller.go
+++ b/pkg/controller/node/node_controller.go
@@ -153,7 +153,8 @@ func isNodeInitializedByCloudControllerManager(node *corev1.Node) bool {
        }
        // Look for the well known taint. If it can be found, then the node is not initialized.
        for _, taint := range node.Spec.Taints {
-               if taint.Key == cloudproviderapi.TaintExternalCloudProvider {
+               if taint.Key == cloudproviderapi.TaintExternalCloudProvider &&
+                       taint.Effect == corev1.TaintEffectNoSchedule {
                        return false
                }
        }
diff --git a/pkg/controller/node/node_controller_test.go b/pkg/controller/node/node_controller_test.go
index be11082..3827d40 100644
--- a/pkg/controller/node/node_controller_test.go
+++ b/pkg/controller/node/node_controller_test.go
@@ -36,6 +36,25 @@ func TestIsNodeInitializedByCloudControllerManager(t *testing.T) {
                        },
                        isInitialized: false,
                },
+               {
+                       node: &corev1.Node{
+                               Spec: corev1.NodeSpec{
+                                       Taints: []corev1.Taint{
+                                               {
+                                                       Key:    "some-other-taint",
+                                                       Value:  "true",
+                                                       Effect: corev1.TaintEffectNoSchedule,
+                                               },
+                                               {
+                                                       Key:    cloudproviderapi.TaintExternalCloudProvider,
+                                                       Value:  "true",
+                                                       Effect: corev1.TaintEffectPreferNoSchedule,
+                                               },
+                                       },
+                               },
+                       },
+                       isInitialized: true,
+               },
                {
                        node: &corev1.Node{
                                Spec: corev1.NodeSpec{
~~~